### PR TITLE
DM-37861: reject new messages with invalid obs_id

### PR DESCRIPTION
**Pull request by r-owen**

### CODE REVIEW

File: alembic/env.py [hunk 1]
```diff
@@ -30,11 +32,12 @@
 # ... etc.
 
 
-def do_run_migrations(connection):
+def do_run_migrations(connection: engine.base.Connection) -> None:
     """Run a migration given an async connection.
 
     A helper function used by run_migrations_online.
     """
+    print("connection is an ", type(connection))
```
Comment (timj) [inline alembic/env.py:40]: Do you mean this print to be here (and why isn't it a log message if it is meant to be here)

File: src/exposurelog/routers/add_message.py [hunk 2]
```diff
@@ -4,19 +4,22 @@
...
 import sqlalchemy as sa
 
 from ..message import ExposureFlag, Message
 from ..shared_state import SharedState, get_shared_state
+from ..utils import current_date_and_day_obs
 from .normalize_tags import TAG_DESCRIPTION, normalize_tags
 
 router = fastapi.APIRouter()
 
+OBSID_REGEX = re.compile(r"[A-Z][A-Z]_[A-Z]_(\d\d\d\d\d\d\d\d)_(\d\d\d\d\d\d)")
```
Comment (timj) [inline src/exposurelog/routers/add_message.py:21]: Do you want a `$` at the end of the regex (and `^` at start) to ensure that you are matching a complete obsid? Consider `r"^[A-Z][A-Z]_[A-Z]_(\d{8})_(\d{6})$"` rather than repeating the `\d`.

File: src/exposurelog/routers/add_message.py [hunk 3]
```diff
@@ -128,11 +132,53 @@ async def add_message(
...
+    """
+    match = OBSID_REGEX.fullmatch(obs_id)
+    if match is None:
+        raise fastapi.HTTPException(
+            status_code=http.HTTPStatus.BAD_REQUEST,
+            detail=f"Invalid {obs_id=}",
+        )
+    obs_id_str = match.groups()[0]
+    if (
+        int(obs_id_str) < current_day_obs - 1
```
Comment (timj) [inline src/exposurelog/routers/add_message.py:167]: if you aren't going to call `int` on the line above how about using `(obs_id_int := int(obs_id_str))` here so you aren't calling `int` on the same thing twice in the same `if`?

File: src/exposurelog/routers/add_message.py [hunk 4]
```diff
@@ -162,7 +208,8 @@ def get_day_obs_from_registries(
                 registry.queryDimensionRecords("exposure", where=query_str)
             )
             if records:
-                return records[0].day_obs
+                print(f"*** {len(records)=}; {type(records[0])=}")
```
Comment (timj) [inline src/exposurelog/routers/add_message.py:211]: not a log message?

File: src/exposurelog/utils.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+__all__ = ["current_date_and_day_obs"]
+
+import astropy.time
+import astropy.units
+
+
+def current_date_and_day_obs() -> tuple[astropy.time.Time, int]:
+    """Return the current date and day_obs."""
+    current_date = astropy.time.Time.now()
+    day_obs_full = current_date.tai - 12 * astropy.units.hr
```
Comment (timj) [inline src/exposurelog/utils.py:10]: Or put the hour offset into a `TimeDelta` object and pull it out as a constant so it doesn't have to be recreated each time?

File: src/exposurelog/utils.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+__all__ = ["current_date_and_day_obs"]
+
+import astropy.time
+import astropy.units
+
+
+def current_date_and_day_obs() -> tuple[astropy.time.Time, int]:
+    """Return the current date and day_obs."""
```
Comment (timj) [inline src/exposurelog/utils.py:8]: Maybe add a note here saying that this matches the LSST day_obs definition of offsetting UTC by 12 hours?

File: tests/test_alembic_migration.py [hunk 7]
```diff
@@ -65,12 +64,15 @@ async def get_column_info(
...
-    def _impl(connection: Connection) -> list[str]:
+    # Note: the return type of the elements of the list
+    # is too complicated to bother trying to annotate.
+    def _impl(connection: Connection) -> list[typing.Any]:
         """Synchronous implementation.
 
         Inspect does not work with an async connection
         """
         inspector = inspect(connection)
+        # Cast the return value to a list to make mypy happy.
```
Comment (timj) [inline tests/test_alembic_migration.py:75]: The line below doesn't look like a cast.

**Review by timj — APPROVED** [review]
Looks good. Some minor comments and suggestions.

File: alembic/env.py (hunk 1 above)
Comment (r-owen) [reply alembic/env.py:40]: Good catch. It was meant to be temporary to figure out mypy annotation. I'll remove it.

File: src/exposurelog/routers/add_message.py (hunk 2 above)
Comment (r-owen) [reply src/exposurelog/routers/add_message.py:21]: I'm using `fullmatch` so there is no need.

File: src/exposurelog/routers/add_message.py (hunk 3 above)
Comment (r-owen) [reply src/exposurelog/routers/add_message.py:167]: Good catch. Also the variable name is misleading; this is a string version of day_obs, not obs_id. I'll just create day_obs as an integer right away (that is safe because the regex matched).

File: src/exposurelog/utils.py (hunk 5 above)
Comment (r-owen) [reply src/exposurelog/utils.py:10]: OK

File: tests/test_alembic_migration.py (hunk 7 above)
Comment (r-owen) [reply tests/test_alembic_migration.py:75]: Good catch. Old message from trying to make mypy happy. Deleted.

File: src/exposurelog/utils.py (hunk 6 above)
Comment (r-owen) [reply src/exposurelog/utils.py:8]: I sure hope that is TAI - 12 hours, but I'll ask on Slack.

File: src/exposurelog/routers/add_message.py (hunk 4 above)
Comment (r-owen) [reply src/exposurelog/routers/add_message.py:211]: More temporary mypy diagnostics. I'll delete.

File: requirements/main.in [hunk 8]
```diff
@@ -5,16 +5,22 @@
...
 #     make update-deps
 
-# prevent a resolution conflict
-# uvicorn 0.19 wants h11 0.14.0, but httpcore 0.15.0 (which is ancient, but demanded by httpx)
-# demands h11 0.12.0
-h11==0.12.0
-alembic~=1.8
-astropy~=5.1
-asyncpg~=0.26
+# Resolve a dependency issue by forcing a version of packaging
```
Comment (wvreeven) [inline requirements/main.in:8]: Very minor comment, but you're mixing comments that don't (like this line) and that do (like line 18) end in a `.`. Please pick one and stick to that.

File: requirements/dev.txt [hunk 9]
```diff
@@ -1,18 +1,18 @@
 #
-# This file is autogenerated by pip-compile with python 3.10
-# To update, run:
+# This file is autogenerated by pip-compile with Python 3.10
```
Comment (wvreeven) [inline requirements/dev.txt:2]: If this file is autogenerated, why is it stored in the repo? Is there a way to generate it when it is needed, e.g. as an additional build step in the CI and conda jobs? Same question for `main.txt`.

File: src/exposurelog/routers/add_message.py (hunk 2 above)
Comment (wvreeven) [reply src/exposurelog/routers/add_message.py:21]: Please add an inline comment to indicate that the intended use is with `fullmatch` so that's clear to anyone reading the code.

File: src/exposurelog/routers/add_message.py [hunk 10]
```diff
@@ -128,44 +126,60 @@ async def add_message(
...
     Parameters
     ----------
     registries
         One or more data registries.
         They are searched in order.
-    obs_id
-        Observation ID.
     instrument
         Instrument name.
+    obs_id
```
Comment (wvreeven) [inline src/exposurelog/routers/add_message.py:143]: Why the change of order in the doc string and not in the argument list itself? Also, please add the expected types to the doc strings of all parameters.

File: src/exposurelog/testutils.py [hunk 11]
```diff
@@ -391,6 +392,23 @@ def random_date(precision: int = 0) -> datetime.datetime:
     ).datetime
 
 
+def random_obs_id() -> str:
+    """Return a random obs_id.
+
+    The format is aa_a_dddddddd_dddddd, where:
+
+    * a is an uppercase letter
```
Comment (wvreeven) [inline src/exposurelog/testutils.py:400]: Perhaps it would be clearer to describe the format as `AA_A_dddddddd_dddddd` so from looking at that format string it already is suggested that A is an uppercase letter? At first I got the impression that `aa_` are lower case and only `_a_` is uppercase.

**Review by wvreeven — APPROVED** [review]
Looks good to me. I left a few minor comments for you to consider.

File: requirements/main.in (hunk 8 above)
Comment (timj) [reply requirements/main.in:8]: Standard python style is to write full sentences with period.

File: requirements/dev.txt (hunk 9 above)
Comment (r-owen) [reply requirements/dev.txt:2]: A given version of the input files can easily lead to broken software as dependencies evolve. One of the harder parts of making any update to this package is dealing with this breakage.

Thus I consider it crucial to commit the generated files, because this pins the exact versions of software that the service will use, leaving the code in a known-to-be-working state.

File: src/exposurelog/routers/add_message.py (hunk 10 above)
Comment (r-owen) [reply src/exposurelog/routers/add_message.py]: Good point. I'll change it everywhere.
